### PR TITLE
522: CI/CD Workflow

### PR DESCRIPTION
**Pull request by JonasAls89**

### CODE REVIEW

File: training/040_projects_and_infrastructure/030_Novice.rst [hunk 1]
```diff
@@ -73,24 +73,91 @@ The last entry in the above example ``"namespaced_identifiers": true`` is partic
...
+
+  The CI/CD Workflow...
+
+  - is the preferred method to work on a Sesam project
+  - needs a configured Sesam CLI and a GIT software to operate
+  - is tightly coupled with the skeletal structure of the node config
+  - ensures that incremental changes can be implemented in an agile manner
+  - minimizes risk of erroneous deployments damaging an ecosystem
+
+Building upon the knowledge you just acquired reading the above section :ref:`node-config-4-2` is should now be obvious to you that Sesam advocate working in a CI/CD workflow.
```
Comment (ga-hegsvold) [inline training/040_projects_and_infrastructure/030_Novice.rst:89]: > is should now be obvious to you

I think we should be careful with using words such as 'obvious'. If it indeed was obvious, would we even need to mention it?

Oh, and a typo: "is" --> "it"

File: training/040_projects_and_infrastructure/030_Novice.rst [hunk 2]
```diff
@@ -73,24 +73,91 @@ The last entry in the above example ``"namespaced_identifiers": true`` is partic
...
 
-Update last seen
+The information relating to a Sesam node are the properties ``SESAM_API_URL`` and ``JWT``.
+These properties allow you to connect to a given Sesam node's API.
 
-reset to end
+With regards to the node config, the properties ``GIT_REPO``, ``SYNC_ROOT`` and ``VARIABLES_FILE_PATH`` are all related to the skeletal structure of the node config.
+The ``GIT_REPO`` is equivalent to your root project folder,
+whilst the ``SYNC_ROOT`` is equivalent to the ``node/`` folder whilst finally the ``VARIABLES_FILE_PATH`` defines which of the variables files,
+placed in the ``variables/`` folder, should be used when the Github Autodeployer automatically uploads an updated node config to your Sesam node as you push changes via the Sesam CLI towards GitHub.
```
Comment (ga-hegsvold) [inline training/040_projects_and_infrastructure/030_Novice.rst:141]: This is a very long sentence.
When describing different parameters and similar things, as a general rule I recommend we give each a separate paragraph. This makes it much easier for the reader to distinguish between the different things being described and to find what is relevant at a glance.

File: training/040_projects_and_infrastructure/030_Novice.rst [hunk 3]
```diff
@@ -73,24 +73,91 @@ The last entry in the above example ``"namespaced_identifiers": true`` is partic
...
+From the above system configuration we will now focus on the ``environment`` dictionary part.
+This part contains information that relates to a given Sesam node in addition to the folder structure in the used node config, as outlined in :ref:`node-config-4-2`.
 
-Update last seen
+The information relating to a Sesam node are the properties ``SESAM_API_URL`` and ``JWT``.
+These properties allow you to connect to a given Sesam node's API.
 
-reset to end
+With regards to the node config, the properties ``GIT_REPO``, ``SYNC_ROOT`` and ``VARIABLES_FILE_PATH`` are all related to the skeletal structure of the node config.
+The ``GIT_REPO`` is equivalent to your root project folder,
```
Comment (ga-hegsvold) [inline training/040_projects_and_infrastructure/030_Novice.rst:139]: ``GIT_REPO`` must contain the link to the git repo where the project's Sesam configuration resides.

File: training/040_projects_and_infrastructure/030_Novice.rst (hunk 2 above)
Comment (ga-hegsvold) [inline training/040_projects_and_infrastructure/030_Novice.rst:141]: > placed in the ``variables/`` folder,

This is redundant. As we can se from the example above, the path to the variables file must be specified, so it does not have to be in the `variables/` folder.

Comment (ga-hegsvold) [inline training/040_projects_and_infrastructure/030_Novice.rst:141]: > as you push changes via the Sesam CLI towards GitHub

I don't think this is relevant when describing how the autodeployer works. Changes can be pushed to github (or whatever VCS is used) in various ways. The autodeployer does not care how.

**Review by ga-hegsvold — CHANGES_REQUESTED** [review]
How much of the CI/CD setup should this section cover?
The autodeployer is mostly explained, which is good, but not sufficient for the reader to set up CI/CD by themselves.
Thoughts?

It's might also be worth point out that CI/CD can be set up in multiple ways for a Sesam project, but what is described here are Sesam's recommendations.

**Comment by JonasAls89** [comment]
> How much of the CI/CD setup should this section cover? The autodeployer is mostly explained, which is good, but not sufficient for the reader to set up CI/CD by themselves. Thoughts?
> 
> It's might also be worth point out that CI/CD can be set up in multiple ways for a Sesam project, but what is described here are Sesam's recommendations.

I'm not completely sure how to answer/resolve the above. As I link in the section to Beginner -> Sesam CLI I think the theory for setting up the CI/CD workflow in Sesam is covered. Don't you agree?

File: training/040_projects_and_infrastructure/030_Novice.rst (hunk 1 above)
Comment (JonasAls89) [reply training/040_projects_and_infrastructure/030_Novice.rst]: implemented

File: training/040_projects_and_infrastructure/030_Novice.rst (hunk 2 above)
Comment (JonasAls89) [reply training/040_projects_and_infrastructure/030_Novice.rst]: split it up as suggested.

File: training/040_projects_and_infrastructure/030_Novice.rst (hunk 3 above)
Comment (JonasAls89) [reply training/040_projects_and_infrastructure/030_Novice.rst]: implemented

File: training/040_projects_and_infrastructure/030_Novice.rst (hunk 2 above)
Comment (JonasAls89) [reply training/040_projects_and_infrastructure/030_Novice.rst]: removed

Comment (JonasAls89) [reply training/040_projects_and_infrastructure/030_Novice.rst]: removed

**Review by gabriellsesam — CHANGES_REQUESTED** [review]
Minor change; we should explain what CI/CD is the abbreviation for  (Continuous integration/ Continuous  deployment)

**Comment by JonasAls89** [comment]
> Minor change; we should explain what CI/CD is the abbreviation for (Continuous integration/ Continuous deployment)

This is explained in previous section. :ref:sesam-cli-4-1, therefore I think it would be redundant to explain again?

**Review by gabriellsesam — APPROVED** [review]
> > Minor change; we should explain what CI/CD is the abbreviation for (Continuous integration/ Continuous deployment)
> 
> This is explained in previous section. :ref:sesam-cli-4-1, therefore I think it would be redundant to explain again?

Agreed

File: training/040_projects_and_infrastructure/030_Novice.rst [hunk 4]
```diff
@@ -73,24 +73,90 @@ The last entry in the above example ``"namespaced_identifiers": true`` is partic
...
+.. The following statement might be assuming a bit too much, but at least if you want continuous integration and deployment in a Sesam project, this is our recommended way of doing it.
+.. I'm adding the comment here instead of in GitHub to presist it in case we want to rephrase later.
+
+As such, you will want your Sesam project work to be continously integrated and continously deployed.
+
+This ensures that incremental changes can be implemented in an agile manner and among other things,
+eases the way in which teams can work together whilst also minimizing risk of erroneous deployments damaging an ecosystem.
+
+To implement said workflow, Sesam has developed a microservice named `Github Autodeployer <https://github.com/sesam-community/github-autodeployer>`_.
+This microservice connects to the GitHub API and integrates with a node config, as described in :ref:`node-config-4-2`, shared via GitHub.
```
Comment (ga-hegsvold) [inline training/040_projects_and_infrastructure/030_Novice.rst:103]: Reading this sentence it sounds like the referenced section (node-config-4-2) explains how the GitHub API and integrates with a node config. That is not really the case, is it? The referenced section explains a typical folder structure for a node config.
I think this reference can be removed.

**Review by ga-hegsvold — CHANGES_REQUESTED** [review]

**Review by ga-hegsvold — APPROVED** [review]

**Pull request merged**